ffi/src/lib.rs

```
@@ -329,6 +329,81 @@ fn propose_on_db(
     Ok(proposal_id)
 }
 
+/// Proposes a batch of operations to the database on top of an existing proposal.
+///
+/// # Arguments
+///
+/// * `db` - The database handle returned by `open_db`
+/// * `proposal_id` - The ID of the proposal to propose on
+/// * `nkeys` - The number of key-value pairs to put
+/// * `values` - A pointer to an array of `KeyValue` structs
+///
+/// # Returns
+///
+/// A `Value` containing {id, nil} if creating the proposal succeeded.
+/// A `Value` containing {0, "error message"} if creating the proposal failed.
+///
+/// # Safety
+///
+/// This function is unsafe because it dereferences raw pointers.
+/// The caller must:
+///  * ensure that `db` is a valid pointer returned by `open_db`
+///  * ensure that `values` is a valid pointer and that it points to an array of `KeyValue` structs of length `nkeys`.
+///  * ensure that the `Value` fields of the `KeyValue` structs are valid pointers.
+///
+#[unsafe(no_mangle)]
+pub unsafe extern "C" fn fwd_propose_on_proposal(
+    db: *const DatabaseHandle,
+    proposal_id: ProposalId,
+    nkeys: usize,
+    values: *const KeyValue,
+) -> Value {
+    // Note: the id is guaranteed to be non-zero
+    // because we use an atomic counter that starts at 1.
+    propose_on_proposal(db, proposal_id, nkeys, values)
+        .map(|id| id.into())
+        .unwrap_or_else(|e| e.into())
+}
+
+/// Internal call for `fwd_propose_on_proposal` to remove error handling from the C API
+#[doc(hidden)]
+fn propose_on_proposal(
+    db: *const DatabaseHandle,
+    proposal_id: ProposalId,
+    nkeys: usize,
+    values: *const KeyValue,
+) -> Result<ProposalId, String> {
+    let db = unsafe { db.as_ref() }.ok_or_else(|| String::from("db should be non-null"))?;
+
+    // Get proposal from ID.
+    // We need write access to add the proposal after we create it.
+    let mut proposals = db.proposals.write().unwrap();
```

Curious as to why this unwrap was accepted by clippy. If you want to keep this here, the function documentation needs a Panics section, but it might be better to return an error if the proposal lock is poisoned.

---

This lock is held a really long time, preventing parallel reads from any proposals. It seems like we should demote this to a read lock here and get a new write lock later.

---

No you're right - I should return an error. For the map lock, that's more confusing. I was worried about that, but I didn't want to hold a reference to the proposal after the lock went out of scope, and it seemed weird to arbitrarily scope the lock. Is that the correct way to do it?

---

I figured out why we aren't getting warnings. You should avoid unwrap. It's really designed for prototyping. If you are 100% sure there is no way the unwrap can fail, then use `expect`.

In this case, the code wants to know what we should do if some thread dies while holding the proposals lock. I think it's okay to die there too, but we should use expect so you get a message saying "poisoned proposal lock" or something like that.